src/script/src/python/pyo3/builtins.rs

```diff
@@ -0,0 +1,335 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use common_function::scalars::{FunctionRef, FUNCTION_REGISTRY};
+use datafusion::arrow::array::{ArrayRef, NullArray};
+use datafusion::physical_plan::expressions;
+use datafusion_expr::ColumnarValue;
+use datafusion_physical_expr::{math_expressions, AggregateExpr};
+use datatypes::vectors::VectorRef;
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use pyo3::types::PyList;
+
+use super::utils::scalar_value_to_py_any;
+use crate::python::ffi_types::utils::all_to_f64;
+use crate::python::ffi_types::PyVector;
+use crate::python::pyo3::dataframe_impl::col;
+use crate::python::pyo3::utils::{
+    columnar_value_to_py_any, try_into_columnar_value, val_to_py_any,
+};
+
+macro_rules! batch_import {
+    ($m: ident, [$($fn_name: ident),*]) => {
+        $($m.add_function(wrap_pyfunction!($fn_name, $m)?)?;)*
+    };
+}
+
+#[pymodule]
+#[pyo3(name = "greptime")]
+pub(crate) fn greptime_builtins(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    batch_import!(
+        m,
+        [
+            col,
+            vector,
+            pow,
+            clip,
+            diff,
+            mean,
+            polyval,
+            argmax,
+            argmin,
+            percentile,
+            scipy_stats_norm_cdf,
+            scipy_stats_norm_pdf,
+            sqrt,
+            sin,
+            cos,
+            tan,
+            asin,
+            acos,
+            atan,
+            floor,
+            ceil,
+            round,
+            trunc,
+            abs,
+            signum,
+            exp,
+            ln,
+            log2,
+            log10,
+            random,
+            approx_distinct,
+            median,
+            approx_percentile_cont,
+            array_agg,
+            avg,
+            correlation,
+            count,
+            covariance,
+            covariance_pop,
+            max,
+            min,
+            stddev,
+            stddev_pop,
+            sum,
+            variance,
+            variance_pop
+        ]
+    );
+    Ok(())
+}
+
+fn eval_func(py: Python<'_>, name: &str, v: &[&PyVector]) -> PyResult<PyVector> {
+    py.allow_threads(|| {
+        let v: Vec<VectorRef> = v.iter().map(|v| v.as_vector_ref()).collect();
+        let func: Option<FunctionRef> = FUNCTION_REGISTRY.get_function(name);
+        let res = match func {
+            Some(f) => f.eval(Default::default(), &v),
+            None => return Err(PyValueError::new_err(format!("Can't find function {name}"))),
+        };
+        match res {
+            Ok(v) => Ok(v.into()),
+            Err(err) => Err(PyValueError::new_err(format!(
+                "Fail to evaluate the function,: {err}"
+            ))),
+        }
+    })
+}
+
+fn eval_aggr_func(py: Python<'_>, name: &str, args: &[&PyVector]) -> PyResult<PyObject> {
+    let res = py.allow_threads(|| {
```

Comment:
When we allow threads, are there any side effects?

Reply:
> When we allow threads, are there any side effects?

`allow_threads` release GIL so other python thread can run, this is useful for cases like this where heavy computaion is happening and python side action is not. So good side effects?